test/rules/use_function_declaration_instead_variable.dart

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N use_function_declaration_instead_variable`
+
+void main() {
+  thisIsGood() { // OK
+    print('this is good');
+  }
+
+  var bad1 = () { // LINT
+    print('this is bad');
+  };
+
+  var bad2 = bad1; // LINT
+
+  bad2 = bad1; // LINT
+
+  bad2 = () { // LINT
```

Comment:
I believe that's reasonable (that the first does not produce a lint while the second does). Again, the point of the guideline (as I understand it) is *not* to prevent assigning a function to a variable, the point is to encourage code that is easier to read. The hypothesis, which I agree with, is that this
```
void bad1() {
  print('this is good');
}
```
is easier to read than
```
var bad1 = () {
 print('this is bad');
}
```
But I don't think we should generate a lint if the variable is re-assigned somewhere later in the function / method, because then the user really *is* trying to create a variable, not just a function.

To put it another way, the guideline is about understanding user intent and helping them accomplish their goal in a better way. In order to do that we have to understand whether they're trying to define a function or a variable.